rpc/jsonrpc/types/types.go

```diff
@@ -77,18 +45,39 @@ var errorCodeString = map[ErrorCode]string{
 // REQUEST
 
 type RPCRequest struct {
-	ID     jsonrpcid
+	id json.RawMessage
+
 	Method string
 	Params json.RawMessage
 }
 
+// NewRequest returns an empty request with the specified ID.
+func NewRequest(id int) RPCRequest {
+	return RPCRequest{id: []byte(strconv.Itoa(id))}
```

Comment:
I'm inclined to think that these should be UUIDs or something to avoid having two clients construct IDs sequentially and have them overlap, but we don't have to go full UUID

Reply:
It doesn't matter—because each request or request batch is part of a single HTTP transaction (even for the websocket), the ID namespace for different clients is completely non-overlapping. The only thing we have to check for is collision within a single HTTP transaction (which we do).

Reply:
(ID collision is only a concern if you have multiple clients talking to the server on the same data stream—but we don't support that at all)

Reply:
ah, sorry, given the name and the package I thought that this was for all requests and not just scoped to a single request.

Reply:
This package does handle all requests, but this type is only the encoding of the JSON-RPC request wrapper to and from JSON. 

On the client side, the client must assign an ID when it issues a call, but as long as the client doesn't reuse the same ID across concurrent requests sent to the service, there is no concern about overlap. The client has to keep track of that, the request doesn't.

On the server side, the server accepts whatever ID the client sent, as long as it is valid (either a string or a number), but otherwise it's just an opaque blob.